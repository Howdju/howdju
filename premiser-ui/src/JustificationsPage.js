import React, {Component} from "react"
import Helmet from "react-helmet"
import {
  Button,
  CircularProgress,
  FontIcon,
  ListItem,
} from "react-md"
import {connect} from "react-redux"
import concat from 'lodash/concat'
import forEach from 'lodash/forEach'
import isEmpty from 'lodash/isEmpty'
import isEqual from 'lodash/isEqual'
import map from 'lodash/map'
import sortBy from "lodash/sortBy"
import split from 'lodash/split'
import toLower from 'lodash/toLower'
import {denormalize} from "normalizr"
import queryString from 'query-string'

import {
  isVerified,
  isDisverified,
  JustificationPolarity,
  JustificationTargetType,
  makeNewTrunkJustification,
} from "howdju-common"

import {
  api,
  editors,
  mapActionCreatorGroupToDispatchToProps,
  ui,
  goto,
  flows,
} from "./actions"
import * as characters from './characters'
import ContextTrail from './ContextTrail'
import JustificationsTree from './JustificationsTree'
import {logger} from "./logger"
import NewJustificationDialog from './NewJustificationDialog'
import {EditorTypes} from "./reducers/editors"
import RootTargetCard from './RootTargetCard'
import t, {
  ADD_JUSTIFICATION_CALL_TO_ACTION,
} from "./texts"
import {
  combineIds,
  combineSuggestionsKeys,
  contextTrailTypeByShortcut,
  describeRootTarget,
  rootTargetNormalizationSchemasByType,
} from './viewModels'

import "./JustificationsPage.scss"


const rootTargetInfoFromProps = (props) => ({
  rootTargetType: props.rootTargetType,
  rootTargetId: props.match.params.rootTargetId,
})

function toContextTrailInfos(contextTrailParam) {
  const contextTrailItemStrings = split(contextTrailParam, ';')
  const contextTrailItems = map(contextTrailItemStrings, (s) => {
    const [typeShortcut, targetId] = split(s, ',')
    return {
      targetType: contextTrailTypeByShortcut[typeShortcut],
      targetId,
    }
  })
  return contextTrailItems
}

const contextTrailInfosFromProps = (props) => {
  const queryParams = queryString.parse(props.location.search)
  const contextTrailParam = queryParams['context-trail']
  return contextTrailParam ? toContextTrailInfos(contextTrailParam) : []
}

/*
 * contextItems/rootTrail/trailRootItems/contextTrail
 * isVerified => isApproved
 *
 * Next steps:
 *
 * implement drill-down
 * types of context items: StatementProposition, Justification, Anchored Statement/Proposition, Justification, Counters
 * StatementTagging
 *
 * counters moving to center when they are expanded
 */

const justificationsPageId = 'justifications-page'

class JustificationsPage extends Component {
  static id = justificationsPageId
  static suggestionsKey = justificationsPageId
  static transientId = 'proposition-justifications-page-proposition'
  static rootTargetEditorId = combineIds(justificationsPageId, 'root-target-editor')
  static newJustificationEditorId = combineIds(justificationsPageId, 'new-justification-editor')

  componentDidMount() {
    const {rootTargetType, rootTargetId} = this.rootTargetInfo()
    this.props.api.fetchRootJustificationTarget(rootTargetType, rootTargetId)

    const contextTrailItems = contextTrailInfosFromProps(this.props)
    if (!isEmpty(contextTrailItems)) {
      this.props.api.fetchJustificationTargets(contextTrailItems)
    }
  }

  componentDidUpdate(prevProps,) {
    const prevRootTargetInfo = rootTargetInfoFromProps(prevProps)
    const rootTargetInfo = this.rootTargetInfo()

    if (!isEqual(rootTargetInfo, prevRootTargetInfo)) {
      this.props.api.fetchRootJustificationTarget(rootTargetInfo.rootTargetType, rootTargetInfo.rootTargetId)
    }

    const prevContextTrailItems = contextTrailInfosFromProps(prevProps)
    const contextTrailItems = contextTrailInfosFromProps(this.props)
    if (!isEqual(contextTrailItems, prevContextTrailItems) && !isEmpty(contextTrailItems)) {
      this.props.api.fetchJustificationTargets(contextTrailItems)
    }
  }

  id = (...args) => combineIds(JustificationsPage.id, ...args)

  suggestionsKey = (...args) => combineSuggestionsKeys(JustificationsPage.suggestionsKey, ...args)

  rootTargetInfo = () => rootTargetInfoFromProps(this.props)

  showNewJustificationDialog = (event, polarity = null) => {
    const {
      rootTargetType,
      rootTargetId,
    } = this.rootTargetInfo()
    const newJustification = makeNewTrunkJustification(rootTargetType, rootTargetId, polarity)
    this.props.editors.beginEdit(EditorTypes.NEW_JUSTIFICATION, JustificationsPage.newJustificationEditorId, newJustification)

    this.props.ui.showNewJustificationDialog()
  }

  showNewPositiveJustificationDialog = (event) => {
    this.showNewJustificationDialog(event, JustificationPolarity.POSITIVE)
  }

  showNewNegativeJustificationDialog = (event) => {
    this.showNewJustificationDialog(event, JustificationPolarity.NEGATIVE)
  }

  saveNewJustification = (event) => {
    event.preventDefault()
    this.props.flows.commitEditThenPutActionOnSuccess(EditorTypes.NEW_JUSTIFICATION, JustificationsPage.newJustificationEditorId, ui.hideNewJustificationDialog())
  }

  cancelNewJustificationDialog = () => {
    this.props.ui.hideNewJustificationDialog()
  }

  render () {
    const {
      rootTargetType,
      rootTarget,
      contextTrailItems,
      sortedJustifications,
      isNewJustificationDialogVisible,
    } = this.props

    const hasJustifications = !isEmpty(sortedJustifications)

    const newContextTrailItems = concat(contextTrailItems, [{
      targetType: rootTargetType,
      target: rootTarget,
    }])

    const rootTargetExtraMenuItems = [
      <ListItem
        primaryText="Add justification"
        key="addJustification"
        leftIcon={<FontIcon>add</FontIcon>}
        onClick={this.showNewJustificationDialog}
      />,
    ]

    return (
      <div id="justifications-page">
        <Helmet>
          <title>
            {rootTarget ?
              describeRootTarget(rootTargetType, rootTarget) :
              `Loading ${describeRootTargetType(rootTargetType)}${characters.ellipsis}`} â€” Howdju
          </title>
        </Helmet>

        <div className="md-grid md-grid--top">
          <ContextTrail trailItems={contextTrailItems} className="md-cell md-cell--12 " />

          <div className="md-cell md-cell--12">

            <RootTargetCard
              id={this.id('root-target')}
              rootTargetType={rootTargetType}
              rootTarget={rootTarget}
              editorId={JustificationsPage.rootTargetEditorId}
              suggestionsKey={this.suggestionsKey('root-target')}
              contextTrailItems={contextTrailItems}
              showJustificationCount={false}
              onShowNewJustificationDialog={this.showNewJustificationDialog}
              extraMenuItems={rootTargetExtraMenuItems}
            />

          </div>

          {!hasJustifications && rootTarget && [
            <div className="md-cell md-cell--12 cell--centered-contents"
                 key="no-justifications-message"
            >
              <div>No justifications.</div>
            </div>,
            <div className="md-cell md-cell--12 cell--centered-contents"
                 key="add-justification-button"
            >
              <Button flat
                      children={t(ADD_JUSTIFICATION_CALL_TO_ACTION)}
                      onClick={this.showNewJustificationDialog}
              />
            </div>
          ]}
        </div>

        {rootTarget ? null : (
          <div className="md-grid md-grid--bottom">
            <div className="md-cell md-cell--12 cell--centered-contents">
              <CircularProgress key="progress" id="justifications-page-progress" />
            </div>
          </div>
        )}

        <JustificationsTree
          justifications={sortedJustifications}
          doShowControls={true}
          doShowJustifications={false}
          isUnCondensed={true}
          showNewPositiveJustificationDialog={this.showNewPositiveJustificationDialog}
          showNewNegativeJustificationDialog={this.showNewNegativeJustificationDialog}
          contextTrailItems={newContextTrailItems}
          className="md-grid--bottom"
        />

        <NewJustificationDialog
          id={this.id('new-justification-dialog')}
          editorId={JustificationsPage.newJustificationEditorId}
          suggestionsKey={this.suggestionsKey('new-justification-dialog')}
          visible={isNewJustificationDialogVisible}
          onCancel={this.cancelNewJustificationDialog}
          onSubmit={this.saveNewJustification}
          onHide={this.cancelNewJustificationDialog}
        />

      </div>
    )
  }
}

function describeRootTargetType(rootTargetType) {
  return toLower(rootTargetType)
}

const sortJustifications = justifications => {
  justifications = sortBy(justifications, j => j.score)
  justifications = sortBy(justifications, j => isDisverified(j) ? 1 : isVerified(j) ? -1 : 0)
  forEach(justifications, j => {
    j.counterJustifications = sortJustifications(j.counterJustifications)
  })
  return justifications
}

const entitiesStoreKeyByJustificationTargetType = {
  [JustificationTargetType.PROPOSITION]: 'propositions',
  [JustificationTargetType.STATEMENT]: 'statements',
  [JustificationTargetType.JUSTIFICATION]: 'justifications',
}

function toContextTrailItem(state, info) {
  const {
    targetType,
    targetId,
  } = info
  const storeKey = entitiesStoreKeyByJustificationTargetType[targetType]
  const target = state.entities[storeKey][targetId]
  return {
    targetType,
    target,
  }
}

const mapStateToProps = (state, ownProps) => {
  const {
    rootTargetType,
    rootTargetId,
  } = rootTargetInfoFromProps(ownProps)

  if (!rootTargetType || !rootTargetId) {
    logger.error(`rootTargetType and rootTargetId are required`)
  }

  const schema = rootTargetNormalizationSchemasByType[rootTargetType]
  const rootTarget = denormalize(rootTargetId, schema, state.entities)

  const contextTrailItems = map(contextTrailInfosFromProps(ownProps), info => toContextTrailItem(state, info) || null)

  const sortedJustifications = rootTarget ? sortJustifications(rootTarget.justifications) : []

  return {
    ...state.ui.justificationsPage,
    rootTargetType,
    rootTarget,
    contextTrailItems,
    sortedJustifications,
  }
}

export default connect(mapStateToProps, mapActionCreatorGroupToDispatchToProps({
  api,
  ui,
  editors,
  goto,
  flows,
}))(JustificationsPage)
