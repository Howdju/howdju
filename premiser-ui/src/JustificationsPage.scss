@import '~react-md/src/scss/react-md';
@import './flatui-colors';
@import './colors';
@import './grid';
@import './components';

$gradient-start-opacity: .6;
$gradient-point: 25%;
$slide-size: $md-btn-icon-size;

// All must be less than $md-toolbar-z-index (default 15) or else they will float above the toolbar upon scrolling
// ($md-menu-z-index is 10)
$gradient-z-index: 5;
$over-gradient-z-index: 10;

.context-menu--proposition {
  z-index: $over-gradient-z-index;
  .md-list--menu {
    z-index: $over-gradient-z-index + 1;
  }
}

ul.proposition-trail {
  padding-left: 35px;
  li {
    @include material-icon-list-item("replay");
    margin: 5px;
  }
}

.context-menu {
  float: right;
}
