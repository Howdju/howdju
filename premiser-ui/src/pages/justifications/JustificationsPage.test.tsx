import React from "react";
import { rest } from "msw";
import { screen } from "@testing-library/react";
import { createMemoryHistory } from "history";
import moment from "moment";
import { merge } from "lodash";

import {
  CreateJustification,
  httpStatusCodes,
  JustificationOut,
  PropositionRef,
  toSlug,
} from "howdju-common";
import {
  InferRequestBody,
  InferResponseBody,
  serviceRoutes,
} from "howdju-service-routes";

import {
  makeRouteComponentProps,
  renderWithProviders,
  setupUserEvent,
  withFakeTimers,
  withMockServer,
  withStaticFromNowMoment,
} from "@/testUtils";
import JustificationsPage from "./JustificationsPage";

withFakeTimers();
const server = withMockServer();

const created = moment("2023-01-12T08:23:00");

describe("JustificationsPage", () => {
  withStaticFromNowMoment("2023-01-12T20:14:00");

  test("Shows a justified proposition", async () => {
    // Arrange
    const justifications: JustificationOut[] = [];
    const proposition = {
      ...PropositionRef.parse({ id: "1" }),
      created,
      text: "the-proposition-text",
      justifications,
    };
    const response: InferResponseBody<typeof serviceRoutes.readProposition> = {
      proposition,
    };

    server.use(
      rest.get(
        `http://localhost/propositions/${proposition.id}`,
        (_req, res, ctx) => {
          return res(ctx.status(httpStatusCodes.OK), ctx.json(response));
        }
      )
    );

    const history = createMemoryHistory();
    const { location, match } = makeRouteComponentProps(
      // TODO(196) convert routesById to routePropsById so that we can get path like `routesById["proposition"].path`
      "p/:rootTargetId/:slug",
      {
        pathParams: {
          rootTargetId: proposition.id,
          slug: toSlug(proposition.text),
        },
      }
    );

    // Act
    const { container } = renderWithProviders(
      <JustificationsPage
        rootTargetType={"PROPOSITION"}
        history={history}
        location={location}
        match={match}
      />,
      { history }
    );

    // Assert
    expect(await screen.findByText(proposition.text)).toBeInTheDocument();
    jest.runAllTimers();
    expect(container).toMatchSnapshot();
  });

  test("Can edit a proposition root target", async () => {
    // Arrange
    const user = setupUserEvent();

    const justifications: JustificationOut[] = [];
    const proposition = {
      ...PropositionRef.parse({ id: "1" }),
      created,
      text: "the-proposition-text",
      justifications,
    };
    const response: InferResponseBody<typeof serviceRoutes.readProposition> = {
      proposition,
    };

    server.use(
      rest.get(
        `http://localhost/propositions/${proposition.id}`,
        (_req, res, ctx) => {
          return res(ctx.status(httpStatusCodes.OK), ctx.json(response));
        }
      ),
      rest.get(`http://localhost/search-propositions`, (_req, res, ctx) => {
        const response: InferResponseBody<
          typeof serviceRoutes.searchPropositions
        > = { propositions: [] };
        return res(ctx.status(httpStatusCodes.OK), ctx.json(response));
      }),
      rest.put(
        `http://localhost/propositions/${proposition.id}`,
        async (req, res, ctx) => {
          const { proposition } = await req.json<
            InferRequestBody<typeof serviceRoutes.updateProposition>
          >();
          const response: InferResponseBody<
            typeof serviceRoutes.updateProposition
          > = { proposition };
          return res(ctx.status(httpStatusCodes.OK), ctx.json(response));
        }
      )
    );

    const history = createMemoryHistory();
    const { location, match } = makeRouteComponentProps(
      // TODO(196) convert routesById to routePropsById so that we can get path like `routesById["proposition"].path`
      "p/:rootTargetId/:slug",
      {
        pathParams: {
          rootTargetId: proposition.id,
          slug: toSlug(proposition.text),
        },
      }
    );

    renderWithProviders(
      <JustificationsPage
        rootTargetType={"PROPOSITION"}
        history={history}
        location={location}
        match={match}
      />,
      { history }
    );
    const updatedText = "The updated text";

    // Act
    await user.click(screen.getByRole("button", { description: "Actions" }));
    await user.click(screen.getByRole("button", { name: /Edit/i }));
    const propositionTextInput = screen.getByLabelText(/Text/i);
    await user.clear(propositionTextInput);
    await user.type(propositionTextInput, updatedText);
    await user.click(screen.getByRole("button", { name: "Save" }));

    // Assert
    jest.runAllTimers();
    expect(await screen.findByText(updatedText)).toBeInTheDocument();
  });

  test("Can add a compound-based justification", async () => {
    // Arrange

    const user = setupUserEvent();
    const justifications: JustificationOut[] = [];
    const rootTargetId = "1";
    const proposition = {
      ...PropositionRef.parse({ id: rootTargetId }),
      created,
      text: "the-proposition-text",
      justifications,
    };

    const basisText1 = "The proposition text 1";
    const basisText2 = "The proposition text 2";
    const justification: CreateJustification = {
      target: {
        type: "PROPOSITION",
        entity: proposition,
      },
      polarity: "POSITIVE",
      basis: {
        type: "PROPOSITION_COMPOUND",
        entity: {
          atoms: [
            { entity: { text: basisText1 } },
            { entity: { text: basisText2 } },
          ],
        },
      },
    };
    server.use(
      rest.get(
        `http://localhost/propositions/${proposition.id}`,
        (_req, res, ctx) => {
          const response: InferResponseBody<
            typeof serviceRoutes.readProposition
          > = {
            proposition,
          };
          return res(ctx.status(httpStatusCodes.OK), ctx.json(response));
        }
      ),
      rest.get(`http://localhost/search-propositions`, (_req, res, ctx) => {
        const response: InferResponseBody<
          typeof serviceRoutes.searchPropositions
        > = { propositions: [] };
        return res(ctx.status(httpStatusCodes.OK), ctx.json(response));
      }),
      rest.post(`http://localhost/justifications`, async (_req, res, ctx) => {
        const response: InferResponseBody<
          typeof serviceRoutes.createJustification
        > = {
          isExtant: false,
          justification: merge({}, justification, {
            id: "1582",
            target: { entity: { id: rootTargetId } },
            basis: {
              entity: {
                id: "910",
                atoms: [{ entity: { id: "911" } }, { entity: { id: "912" } }],
              },
            },
            rootTargetType: "PROPOSITION",
            rootTarget: { id: rootTargetId },
          }) as JustificationOut,
        };
        return res(ctx.status(httpStatusCodes.OK), ctx.json(response));
      })
    );

    const history = createMemoryHistory();
    const { location, match } = makeRouteComponentProps(
      // TODO(196) convert routesById to routePropsById so that we can get path like `routesById["proposition"].path`
      "p/:rootTargetId/:slug",
      {
        pathParams: {
          rootTargetId: proposition.id,
          slug: toSlug(proposition.text),
        },
      }
    );

    const { container } = renderWithProviders(
      <JustificationsPage
        rootTargetType={"PROPOSITION"}
        history={history}
        location={location}
        match={match}
      />,
      { history }
    );

    // Act
    await user.click(screen.getByRole("button", { name: /Add one now/i }));
    await user.type(screen.getByLabelText(/Text/i), basisText1);
    await user.click(screen.getByRole("button", { description: /add atom/i }));
    await user.type(screen.getAllByLabelText(/Text/i)[0], basisText2);
    await user.click(screen.getByRole("button", { name: /Create/i }));

    // Assert
    expect(await screen.findByText(basisText1)).toBeInTheDocument();
    expect(await screen.findByText(basisText2)).toBeInTheDocument();

    jest.runAllTimers();
    expect(container).toMatchSnapshot();
  });
});
