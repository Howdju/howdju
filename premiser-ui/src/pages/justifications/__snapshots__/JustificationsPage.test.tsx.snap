// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JustificationsPage Shows a justified proposition 1`] = `
<div>
  <div
    id="justifications-page"
  >
    <div
      class="md-grid md-grid--top"
    >
      <div />
      <div
        class="md-cell md-cell--12"
      >
        <div
          class=""
        >
          <div
            class="root-target-background"
          >
            <div
              class="rmd-card rmd-card--shadowed root-target-card"
              style="width: 100%;"
            >
              <div
                class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin rmd-card__content--secondary root-target-card-contents"
              >
                <div
                  class="entity-viewer"
                >
                  <div
                    class="entity-viewer--header"
                  >
                    <a
                      href="/p/1/the-proposition-text"
                    >
                      <i
                        class="md-icon material-icons"
                        role="presentation"
                        title="Proposition"
                      >
                        short_text
                      </i>
                    </a>
                  </div>
                  <div
                    class="entity-viewer--entity"
                  >
                    <div
                      class="proposition-viewer"
                      id="justifications-page--root-target--proposition-entity-viewer"
                    >
                      <div
                        class="proposition-viewer"
                      >
                        <div
                          class="proposition-text"
                        >
                          <a
                            href="/p/1/the-proposition-text"
                          >
                            the-proposition-text
                          </a>
                        </div>
                        <div
                          class="relation-counts"
                        >
                           
                          <a
                            href="/propositions/1/usages"
                            title="used in 0 justifications"
                          >
                            <span
                              class="material-symbols"
                              style="font-variation-settings: "opsz" 12; font-size: 12px;"
                            >
                              merge_type
                            </span>
                            0
                          </a>
                           
                          <span
                            class="entity-status-text"
                          >
                            <a
                              class="clickable"
                              title="0 appearances"
                            >
                              <span
                                class="material-symbols"
                                style="font-variation-settings: "opsz" 12; font-size: 12px;"
                              >
                                pin_drop
                              </span>
                              0
                            </a>
                          </span>
                        </div>
                        <div>
                          <span
                            class="entity-status-text"
                          >
                            created
                            
                             
                            <span
                              title="12 Jan 2023 4:23:12 PM +00:00"
                            >
                              8 hours ago
                            </span>
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="md-layover md-layover--simplified md-inline-block md-menu-container context-menu context-menu--proposition"
                    id="justifications-page--root-target--menu"
                  >
                    <button
                      class="md-btn md-btn--icon md-pointer--hover md-inline-block hidden"
                      id="justifications-page--root-target--menu-toggle"
                      title="Actions"
                      type="button"
                    >
                      <div
                        class="md-ink-container"
                      />
                      <i
                        class="md-icon material-icons md-text--inherit"
                      >
                        more_vert
                      </i>
                    </button>
                  </div>
                </div>
                <div
                  class="chips-list"
                >
                  <button
                    class="rmd-button rmd-button--icon rmd-button--text-primary"
                    title="Add tag"
                    type="button"
                  >
                    <i
                      aria-hidden="true"
                      class="rmd-icon rmd-icon--font material-icons"
                    >
                      add
                    </i>
                    <span
                      class="rmd-ripple-container"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="md-cell md-cell--12 cell--centered-contents"
      >
        <div>
          No justifications.
        </div>
      </div>
      <div
        class="md-cell md-cell--12 cell--centered-contents"
      >
        <button
          class="rmd-button rmd-button--text rmd-button--text-primary"
          type="button"
        >
          Add one now
          <span
            class="rmd-ripple-container"
          />
        </button>
      </div>
    </div>
    <div
      class="md-grid--bottom md-grid"
    >
      <div
        class="md-cell md-cell--6 md-cell--8-tablet md-cell--4-phone proposition-justifications-justification-trees--positive"
        style="position: relative;"
      />
      <div
        class="md-cell md-cell--6 md-cell--8-tablet md-cell--4-phone proposition-justifications-justification-trees--negative"
        style="position: relative;"
      />
    </div>
  </div>
</div>
`;

exports[`JustificationsPage can add a compound-based justification 1`] = `
<div>
  <div
    id="justifications-page"
  >
    <div
      class="md-grid md-grid--top"
    >
      <div />
      <div
        class="md-cell md-cell--12"
      >
        <div
          class=""
        >
          <div
            class="root-target-background"
          >
            <div
              class="rmd-card rmd-card--shadowed root-target-card"
              style="width: 100%;"
            >
              <div
                class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin rmd-card__content--secondary root-target-card-contents"
              >
                <div
                  class="entity-viewer"
                >
                  <div
                    class="entity-viewer--header"
                  >
                    <a
                      href="/p/1/the-proposition-text"
                    >
                      <i
                        class="md-icon material-icons"
                        role="presentation"
                        title="Proposition"
                      >
                        short_text
                      </i>
                    </a>
                  </div>
                  <div
                    class="entity-viewer--entity"
                  >
                    <div
                      class="proposition-viewer"
                      id="justifications-page--root-target--proposition-entity-viewer"
                    >
                      <div
                        class="proposition-viewer"
                      >
                        <div
                          class="proposition-text"
                        >
                          <a
                            href="/p/1/the-proposition-text"
                          >
                            the-proposition-text
                          </a>
                        </div>
                        <div
                          class="relation-counts"
                        >
                           
                          <a
                            href="/propositions/1/usages"
                            title="used in 0 justifications"
                          >
                            <span
                              class="material-symbols"
                              style="font-variation-settings: "opsz" 12; font-size: 12px;"
                            >
                              merge_type
                            </span>
                            0
                          </a>
                           
                          <span
                            class="entity-status-text"
                          >
                            <a
                              class="clickable"
                              title="0 appearances"
                            >
                              <span
                                class="material-symbols"
                                style="font-variation-settings: "opsz" 12; font-size: 12px;"
                              >
                                pin_drop
                              </span>
                              0
                            </a>
                          </span>
                        </div>
                        <div>
                          <span
                            class="entity-status-text"
                          >
                            created
                            
                             
                            <span
                              title="12 Jan 2023 4:23:12 PM +00:00"
                            >
                              8 hours ago
                            </span>
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="md-layover md-layover--simplified md-inline-block md-menu-container context-menu context-menu--proposition"
                    id="justifications-page--root-target--menu"
                  >
                    <button
                      class="md-btn md-btn--icon md-pointer--hover md-inline-block hidden"
                      id="justifications-page--root-target--menu-toggle"
                      title="Actions"
                      type="button"
                    >
                      <div
                        class="md-ink-container"
                      />
                      <i
                        class="md-icon material-icons md-text--inherit"
                      >
                        more_vert
                      </i>
                    </button>
                  </div>
                </div>
                <div
                  class="chips-list"
                >
                  <button
                    class="rmd-button rmd-button--icon rmd-button--text-primary"
                    title="Add tag"
                    type="button"
                  >
                    <i
                      aria-hidden="true"
                      class="rmd-icon rmd-icon--font material-icons"
                    >
                      add
                    </i>
                    <span
                      class="rmd-ripple-container"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="md-grid--bottom md-grid"
    >
      <div
        class="md-cell md-cell--6 md-cell--8-tablet md-cell--4-phone proposition-justifications-justification-trees--positive"
        style="position: relative;"
      >
        <h2
          class="md-cell md-cell--12"
          style="transition: transform 600ms ease 0ms, opacity 600ms ease 0ms;"
        >
          Supporting Justifications
        </h2>
        <div
          style="transition: transform 615ms ease 50ms, opacity 615ms ease 50ms;"
        >
          <div
            class="justification-tree"
            id="justification-1582-tree"
          >
            <div
              class="md-grid chat-bubble"
            >
              <div
                class="justification-basis-viewer-row"
              >
                <div
                  class="justification-basis-viewer"
                >
                  <ol
                    class="compound-viewer proposition-compound-viewer"
                  >
                    <li
                      class="compound-atom proposition-atom"
                      id="justification-1582-chat-bubble-basis-viewer--proposition-atom-911--list-item"
                    >
                      <div
                        class="entity-viewer"
                      >
                        <div
                          class="entity-viewer--header"
                        >
                          <a
                            href="/p/911/the-proposition-text-1"
                          >
                            <i
                              class="md-icon material-icons"
                              role="presentation"
                              title="Proposition"
                            >
                              short_text
                            </i>
                          </a>
                        </div>
                        <div
                          class="entity-viewer--entity"
                        >
                          <div
                            class="proposition-viewer"
                            id="justification-1582-chat-bubble-basis-viewer--proposition-atom-911--list-item--proposition"
                          >
                            <div
                              class="proposition-viewer"
                            >
                              <div
                                class="proposition-text"
                              >
                                <a
                                  href="/p/911/the-proposition-text-1?context-trail=j,1582,p"
                                >
                                  The proposition text 1
                                </a>
                              </div>
                              <div
                                class="relation-counts"
                              >
                                <a
                                  href="/p/911/the-proposition-text-1?context-trail=j,1582,p"
                                >
                                  <span
                                    title="0 supporting justifications; 0 opposing justifications"
                                  >
                                    <span
                                      class="material-symbols"
                                      style="font-variation-settings: "opsz" 12; font-size: 12px;"
                                    >
                                      merge
                                    </span>
                                    0
                                    +/
                                    0
                                    -
                                  </span>
                                </a>
                                 
                                <a
                                  href="/propositions/911/usages"
                                  title="used in 0 justifications"
                                >
                                  <span
                                    class="material-symbols"
                                    style="font-variation-settings: "opsz" 12; font-size: 12px;"
                                  >
                                    merge_type
                                  </span>
                                  0
                                </a>
                                 
                                <span
                                  class="entity-status-text"
                                >
                                  <a
                                    class="clickable"
                                    title="0 appearances"
                                  >
                                    <span
                                      class="material-symbols"
                                      style="font-variation-settings: "opsz" 12; font-size: 12px;"
                                    >
                                      pin_drop
                                    </span>
                                    0
                                  </a>
                                </span>
                              </div>
                              <div>
                                <span
                                  class="entity-status-text"
                                >
                                  created
                                  
                                   
                                  <span
                                    title=""
                                  />
                                </span>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="compound-atom proposition-atom"
                      id="justification-1582-chat-bubble-basis-viewer--proposition-atom-912--list-item"
                    >
                      <div
                        class="entity-viewer"
                      >
                        <div
                          class="entity-viewer--header"
                        >
                          <a
                            href="/p/912/the-proposition-text-2"
                          >
                            <i
                              class="md-icon material-icons"
                              role="presentation"
                              title="Proposition"
                            >
                              short_text
                            </i>
                          </a>
                        </div>
                        <div
                          class="entity-viewer--entity"
                        >
                          <div
                            class="proposition-viewer"
                            id="justification-1582-chat-bubble-basis-viewer--proposition-atom-912--list-item--proposition"
                          >
                            <div
                              class="proposition-viewer"
                            >
                              <div
                                class="proposition-text"
                              >
                                <a
                                  href="/p/912/the-proposition-text-2?context-trail=j,1582,p"
                                >
                                  The proposition text 2
                                </a>
                              </div>
                              <div
                                class="relation-counts"
                              >
                                <a
                                  href="/p/912/the-proposition-text-2?context-trail=j,1582,p"
                                >
                                  <span
                                    title="0 supporting justifications; 0 opposing justifications"
                                  >
                                    <span
                                      class="material-symbols"
                                      style="font-variation-settings: "opsz" 12; font-size: 12px;"
                                    >
                                      merge
                                    </span>
                                    0
                                    +/
                                    0
                                    -
                                  </span>
                                </a>
                                 
                                <a
                                  href="/propositions/912/usages"
                                  title="used in 0 justifications"
                                >
                                  <span
                                    class="material-symbols"
                                    style="font-variation-settings: "opsz" 12; font-size: 12px;"
                                  >
                                    merge_type
                                  </span>
                                  0
                                </a>
                                 
                                <span
                                  class="entity-status-text"
                                >
                                  <a
                                    class="clickable"
                                    title="0 appearances"
                                  >
                                    <span
                                      class="material-symbols"
                                      style="font-variation-settings: "opsz" 12; font-size: 12px;"
                                    >
                                      pin_drop
                                    </span>
                                    0
                                  </a>
                                </span>
                              </div>
                              <div>
                                <span
                                  class="entity-status-text"
                                >
                                  created
                                  
                                   
                                  <span
                                    title=""
                                  />
                                </span>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </li>
                  </ol>
                </div>
                <div
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="md-layover md-layover--simplified md-inline-block md-menu-container context-menu justification-context-menu"
                  id="justification-1582-context-menu"
                >
                  <button
                    class="md-btn md-btn--icon md-pointer--hover md-inline-block hidden"
                    id="justification-1582-context-menu-toggle"
                    title="Justification actions"
                    type="button"
                  >
                    <div
                      class="md-ink-container"
                    />
                    <i
                      class="md-icon material-icons md-text--inherit"
                    >
                      more_vert
                    </i>
                  </button>
                </div>
              </div>
              <div
                class="md-cell md-cell--12 actions"
              >
                <button
                  class="rmd-button rmd-button--icon rmd-button--text-primary inactive hiding"
                  title="Verify this justification"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="rmd-icon rmd-icon--font material-icons"
                  >
                    thumb_up
                  </i>
                  <span
                    class="rmd-ripple-container"
                  />
                </button>
                <button
                  class="rmd-button rmd-button--icon rmd-button--text-primary inactive hiding"
                  title="Dis-verify this justification"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="rmd-icon rmd-icon--font material-icons"
                  >
                    thumb_down
                  </i>
                  <span
                    class="rmd-ripple-container"
                  />
                </button>
                <button
                  class="rmd-button rmd-button--icon rmd-button--text-primary hiding"
                  title="Counter this justification"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="rmd-icon rmd-icon--font material-icons"
                  >
                    reply
                  </i>
                  <span
                    class="rmd-ripple-container"
                  />
                </button>
                <div
                  class="justification-status-text"
                >
                  <span
                    class="entity-status-text"
                  >
                    created
                    
                     
                    <span
                      title=""
                    />
                  </span>
                </div>
              </div>
              <div
                class="counter-justifications"
              >
                <div
                  style="position: relative;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="md-cell md-cell--6 md-cell--8-tablet md-cell--4-phone proposition-justifications-justification-trees--negative"
        style="position: relative;"
      >
        <h2
          class="md-cell md-cell--12"
          style="transition: transform 600ms ease 0ms, opacity 600ms ease 0ms;"
        >
          Opposing Justifications
        </h2>
        <div
          class="md-cell md-cell--12 cell--centered-contents"
          style="transition: transform 615ms ease 50ms, opacity 615ms ease 50ms;"
        >
          None
        </div>
        <div
          class="md-cell md-cell--12 cell--centered-contents"
          style="transition: transform 630ms ease 100ms, opacity 630ms ease 100ms;"
        >
          <button
            class="rmd-button rmd-button--text rmd-button--text-primary"
            type="button"
          >
            Add one now
            <span
              class="rmd-ripple-container"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
